### DISCUSSION TITLE
To Do List

### DISCUSSION THREAD

**Discussion by Creastery** [Jh123x/cs2102-project#3] [General]
## Priorities:
- [ ] Fix all PK in each table to use single attribute (e.g. sid) instead of a composite FK
- [ ] Populate some sample mock data
- [ ] Identify any duplicate values in tables and see if we need to split up tables more
- [ ] Implement all removed composite FKs constraints into triggers

We also need to mark out the rough changes made to the ERD, like what tables are being joined (e.g. Offerings + Handles) so we can kind of know what tables we are still lacking.

## Schemas:
- [X] Consists (sessions/offerings) - merged into Sessions
- [X] Conducts (instructor/sessions/rooms) - merged into Sessions
- [X] For (payslip/employee) - merged into PaySlips
- [X] Has (courses/offerings) - merged into CourseOfferings
- [X] Handles (administrators/offerings) - merged into CourseOfferings
- [X] In (courses/course_areas) - merged into Courses

https://github.com/Jh123x/cs2102-project/blob/main/SQL/schema/CourseOffering.sql#L12-L13
need to put both param as not null ^

## To Discuss
1. [Registers Table]Customer can register for at most one of its sessions before its registration deadline
2. [Course Offering Table]Status of the course offering. Fully booked: registration > seating capacity. Available: registration < seating capacity
3. [Course Packages/Buy Table] Status of course package. Active: atleast one unused session in package. Partially Active: all sessions redeemed but there is atleast one redeemed session that can be refunded if cancelled. Inactive: all used. 
4. [Course Packages Table]Customer have at most 1 active/partially active package
5. [Cancels Table]Credit card payment cancellation policy: refund 90 % of registered course if cancellation is at least 7 days before the day of the registered session. otherwise: no cancellation
6. [Cancels Table]Redeemed course session cancellation policy: credit an extra course session to customer's course package if cancellation is atleast 7 days before registered session. otherwise: no cancellation
7. [Sessions Table]Instructor cannot be assigned to teach 2 consecutive course session
8. [Sessions Table]Part time instructor cannot teach more than 30 hours per month
10.[Pay_Slips Table]Salary

[TRIGGER] -- Haven't do
1. [Registers Table] Registration Deadline is atleast 10 days before start date

[TRIGGER] -- done 
1. [Sessions Table]Room can be used to conduct at most one course session at any time

## Using Team PostgreSQL Server
`cs2102=> SET search_path TO team80;`

SQL Schema Import Order:
~~~
Rooms.sql
Employees.sql
PaySlips.sql
Customers.sql
CoursePackages.sql
CourseAreas.sql
Courses.sql
CourseOfferings.sql
Specializes.sql
Sessions.sql
OwnsBuysRedeemsRegisters.sql
~~~

i.e. `cat Rooms.sql Employees.sql PaySlips.sql Customers.sql CoursePackages.sql CourseAreas.sql Courses.sql CourseOfferings.sql Specializes.sql Sessions.sql OwnsBuysRedeemsRegisters.sql`

**Comment by daniellimws** [comment]
Replace all `max(id) + 1` for insertions and replace with an awesome postgres syntax that's apparently super clean and concurrent-safe: `insert into table (cols) values (col_values) RETURNING id_column_name_here INTO customer_id_varname_here`

**Reply by Jh123x** [reply]
AHHAHAHHA
